keystone-ioctl.c

```diff
@@ -341,3 +345,19 @@ long keystone_ioctl(struct file *filep, unsigned int cmd, unsigned long arg)
 
   return ret;
 }
+
+int keystone_release(struct inode *inode, struct file *file) {
+  struct keystone_enclave_t* enclave = file->private_data;
+  if (!enclave) {
+    /* If eid is set to the invalid id, then we do not do anything. */
+    return -EINVAL;
+  }
+  pr_info("Releasing enclave: %d\n", enclave->eid);
+  if (enclave->close_on_pexit) {
+    /* We need to send destroy enclave just the eid to close. */
+    struct keystone_ioctl_create_enclave enclp;
+    enclp.eid = enclave->eid;
```

Comment:
I think this will fail to find a valid enclave.
The eid input to the ioctl function is actually user eid, which is different from what enclave data structure has.
Please refer to following code of other ioctls:
```enclave = get_enclave_by_id(ueid);```
Also, wouldn't calling `keystone_destroy_enclave` here cause some double-destroy error?
For example, if the host calls this, and then closes the file, then the `keystone_destroy_enclave` will be called again with error message?

Reply:
I did not know that about the ueid. I just did not want to have to change the function call so I thought I could hack it since it looks like it would be possible how it is being passed in. I have not tested it fully tbh on enclaves which do not call enclave destroy first because I have to mod the sdk in addition to get that to be testable. Also I made sure it is not a double free issue since when an enclave is destroyed, I set the private data to NULL which is why there is the first check. Also since I am calling the destroy function, it will remove the enclave so that you cannot access it. Also I see what you mean by it is a different ID. I guess I will just have to make the destroy function a helper function instead :'(. I will work on that tomorrow.

Reply:
Would it be bad if I added another field which was essentially a bool which  controlled how we tread the arg and the beginning execution of the destroy function?

Reply:
Since `keystone_destroy_enclave` is an ioctl function, it would be better to keep those two functions in a same level of abstraction, instead of making `keystone_release` call `keystone_destroy_enclave`.

I think an easier way to produce cleaner code is to write `keystone_release` just as needed, and then see if which parts are duplicated with `keystone_destroy_enclave`, and pull the part to create an internal function that is used by both of them.

I think for the ioctl, you must check if the enclave id is valid because the input is from user, but for the release_enclave, it seems safe for me to just get eid from enclave struct.

Reply:
Also can you test if destroy_enclave works by inserting some seg fault into the host?
to see if enclave is destroyed when the host dies

Reply:
It has now been fixed.